src/backend/distributed/executor/multi_utility.c

```diff
@@ -2921,6 +2934,126 @@ ErrorIfUnsupportedTruncateStmt(TruncateStmt *truncateStatement)
+ * ProcessTruncateStatement handles distributed locking
+ * of truncated tables before standard utility takes over.
+ *
+ * This is only for distributed serialization of truncate commands.
+ * The function assumes that there is no foreign key relation between
+ * non-distributed and distributed relations.
+ */
+static void
+ProcessTruncateStatement(TruncateStmt *truncateStatement)
+{
+	List *distributedRelationList = NIL;
+
+	ListCell *relationCell = NULL;
+	foreach(relationCell, truncateStatement->relations)
+	{
+		RangeVar *relationRV = (RangeVar *) lfirst(relationCell);
+		Relation relation = heap_openrv(relationRV, NoLock);
+		Oid relationId = RelationGetRelid(relation);
+		DistTableCacheEntry *cacheEntry = NULL;
+		List *referencingTableList = NIL;
+		ListCell *referencingTableCell = NULL;
+
+		if (!IsDistributedTable(relationId))
+		{
+			heap_close(relation, NoLock);
+			continue;
+		}
+
+		if (list_member_oid(distributedRelationList, relationId))
+		{
+			heap_close(relation, NoLock);
+			continue;
+		}
+
+		distributedRelationList = lappend_oid(distributedRelationList, relationId);
+
+		cacheEntry = DistributedTableCacheEntry(relationId);
+		Assert(cacheEntry != NULL);
+
+		referencingTableList = cacheEntry->referencingRelationsViaForeignKey;
+		foreach(referencingTableCell, referencingTableList)
+		{
+			Oid referencingRelationId = lfirst_oid(referencingTableCell);
+			distributedRelationList = list_append_unique_oid(distributedRelationList,
+															 referencingRelationId);
+		}
+
+		heap_close(relation, NoLock);
+	}
+
+	if (distributedRelationList != NIL)
+	{
+		BeginOrContinueCoordinatedTransaction();
+		AcquireDistributedLockOnRelations(distributedRelationList, AccessExclusiveLock,
+										  REMOTE_LOCK_MODE_FOR_TRUNCATE);
+	}
+}
+
+
+/*
+ * AcquireDistributedLockOnRelations acquire a distributed lock on worker nodes
+ * for given list of relations ids. Relation id list and worker node list
+ * sorted so that the lock is acquired in the same order regardless of which
+ * node it was run on. Notice that no lock is acquired on coordinator node.
+ * Both local and remote lock modes are expected to be of same type. Reason
+ * we have both here is that we needed the ENUM for local locking function, and
+ * string representation for remote locking function.
+ *
+ * Notice that the locking functions is sent to all workers regardless of if
+ * it has metadata or not. This is because a worker node only knows itself
+ * and previous workers that has metadata sync turned on. It does not know
+ * about nodes that have metadata sync turned on afterwards.
+ */
+static void
+AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE localLockMode,
+								  char *remoteLockMode)
+{
+	ListCell *relationIdCell = NULL;
+	List *workerNodeList = ActivePrimaryNodeList();
+	relationIdList = SortList(relationIdList, CompareOids);
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodes);
+
+	foreach(relationIdCell, relationIdList)
```

Comment:
Hmm, 
```
foreach relation
      for each worker:
```
could be an expensive loop if there are many foreign keys. What if we do the following:

```
foreach relation:
    build the string

for each worker:
     send the command (or acquire the list locally)
```

I think that would even make the window for distributed deadlocks with concurrent `SELECT`s smaller. Because, with the current commit, if you run a concurrent `SELECT`, we're likely to distributed (or local) deadlocks while waiting for relation locks.

(Or even changing the API for `LOCK_RELATION_IF_EXISTS ` to get a list of relations could be simpler?)